include/libcyphal/transport/contiguous_payload.hpp

```diff
@@ -0,0 +1,109 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_TRANSPORT_CONTIGUOUS_PAYLOAD_HPP_INCLUDED
+#define LIBCYPHAL_TRANSPORT_CONTIGUOUS_PAYLOAD_HPP_INCLUDED
+
+#include "types.hpp"
+#include "libcyphal/types.hpp"
+
+namespace libcyphal
+{
+namespace transport
+{
+
+/// Internal implementation details of the transport layer.
+/// Not supposed to be used directly by the users of the library.
+///
+namespace detail
+{
+
+/// @breaf Makes a contiguous payload from a list of payload fragments.
+///
+/// Has optimization for the case when there is only one non-empty fragment -
+/// in this case there will be no memory allocation and payload copying.
+/// Automatically deallocates memory (if any) when the object is destroyed.
+///
+/// Probably could be deleted when libcanard will start support fragmented payloads (at `canardTxPush`).
+/// See https://github.com/OpenCyphal/libcanard/issues/223
+///
+class ContiguousPayload final
+{
+public:
+    ContiguousPayload(cetl::pmr::memory_resource& mr, const PayloadFragments payload_fragments)
+        : mr_{mr}
+        , payload_{nullptr}
+        , payload_size_{0}
+        , allocated_buffer_{nullptr}
+    {
+        using Fragment = cetl::span<const cetl::byte>;
+
+        // Count fragments skipping empty ones. Also keep tracking of the total payload size
+        // and pointer to the last non-empty fragment (which will be in use for the optimization).
+        //
+        const auto total_non_empty_fragments =
+            std::count_if(payload_fragments.begin(), payload_fragments.end(), [this](const Fragment frag) {
+                if (frag.empty())
+                {
+                    return false;
+                }
+                payload_ = frag.data();
+                payload_size_ += frag.size();
+                return true;
+            });
+
+        if (total_non_empty_fragments > 1)
+        {
+            allocated_buffer_ = static_cast<cetl::byte*>(mr_.allocate(payload_size_));
+            payload_          = allocated_buffer_;
+            if (cetl::byte* const buffer = allocated_buffer_)
+            {
+                std::size_t offset = 0;
+                for (const Fragment frag : payload_fragments)
+                {
+                    std::memmove(&buffer[offset], frag.data(), frag.size());
+                    offset += frag.size();
+                }
+            }
+        }
+    }
+    ContiguousPayload(const ContiguousPayload&)                = delete;
+    ContiguousPayload(ContiguousPayload&&) noexcept            = delete;
+    ContiguousPayload& operator=(const ContiguousPayload&)     = delete;
+    ContiguousPayload& operator=(ContiguousPayload&&) noexcept = delete;
+
+    ~ContiguousPayload()
+    {
+        if (allocated_buffer_ != nullptr)
```

Comment:
There is no way to cover this branch, right? `allocated_buffer_` being nullptr is always undefined behaviour. If so then remove the `if`

Reply:
There is way (with mock memory resource tuned to emulate "out of memory" condition), and there is unit test to cover it: `TEST_F(TestContiguousPayload, ctor_no_memory_error).

Reply:
Also, in the description of PR I attached screenshot of coverage - as you can see almost everything is 100% (lines, functions AND branches). The only exception is `transport.hpp` - it is expected. It will be 100% on next PR where I finished service sessions.